src/QMCDrivers/QMCCostFunctionBase.h

```diff
@@ -250,6 +252,8 @@ class QMCCostFunctionBase : public CostFunctionBase<QMCTraits::RealType>, public
    * is normally the same as OptVariables.
    */
   opt_variables_type OptVariablesForPsi;
+  // unchanged initial checked-in variables
+  opt_variables_type InitVariables;
```

Comment:
it concerns me that this extra InitVariables adds more memory and bookkeeping. For other parameters checked in from the XML input and output to the opt.xml don't need initVariables. Why it is necessary to have InitVariables particularly for CJ.

Reply:
My Jastrow factor contains variables that are not constants (i.e. can be input by the user) yet should not be optimizable (due to parameter redundancies). This is because the Jastrow basis (the "counting functions") are built from a set of normalized gaussians {g_I}: 

C_I(r_i) = (g_I(r_i)) / (\sum_{J} g_J(r_i) } ), 

The gaussian exponents of each g_I are used as optimizable parameters. Dividing by/multiplying by all the counting functions by an arbitrary gaussian function produces the same set of counting functions, and optimizing the all of the gaussian parameters together would therefore produce singular behavior in the linear method. One of the gaussian functions is designated as a 'reference' and should be flagged as not optimizable; requiring the user to use a constant reference gaussian would make the input file both more difficult to read and write.

The only way to write the counting jastrow parameters to the opt.xml files is to read these non-optimized variables from the wavefunction VariableSet. However, variables which are flagged as not optimizable are completely unregistered from both OptVariablesForPsi and OptVariables before this function is called. As a result, I copy the initial variables a single time to another VariableSet before printing so I have their values on hand.

Without doing something like keeping the unoptimized variables in one of the two present VariableSets or moving part of this xml updating function to the WavefunctionComponent class itself, there's not another obvious, simple way to get these values of these non-optimizable variables into the final opt.xml file.

Reply:
I agree with your analysis but I prefer a harder but logically clean route.
I think we need to clean up how to output an input xml block or hdf5 tree for WavefunctionComponent and SPOSet.

In "./molecules/H4_ae/qmc_cjs_ref/H4_cjs_ref.wfs.xml" is the function g0 fixed?
By looking at the input file, I saw the first xml seems treated fixed implicitly.
I would expect a control key optimize="yes"/"no"
I remember in Jastrow factors, when optimize="no", the code doesn't optimize certain correlation pairs and the opt.xml can still print them. Could you have a look of how the two body Jastrow xml was handled?